python_modules/dagster/dagster/_core/storage/root.py

```diff
@@ -49,3 +50,25 @@ def from_config_value(
     @classmethod
     def config_type(cls) -> UserConfigSchema:
         return {"base_dir": StringSource}
+
+    def dispose(self):
+        pass
+
+
+class TemporaryLocalArtifactStorage(LocalArtifactStorage):
+    """Used by ephemeral DagsterInstances, defers directory creation til
+    access since many uses of ephemeral instance do not require artifact directory.
+    """
+
+    def __init__(self):
+        self._temp_dir = None
+
+    @property
+    def base_dir(self):
+        if self._temp_dir is None:
+            self._temp_dir = TemporaryDirectory()
+        return self._temp_dir.name
+
+    def dispose(self):
+        if self._temp_dir:
+            self._temp_dir.cleanup()
```

Comment:
this could use an ExitStack too but cleanup() maps to __exit__ i imagine?

Reply:
yep - this is a more direct variant